Add a stable TORCH_LIBRARY to C shim

janeyx99 opened the pull request:
This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #148832
* __->__ #148124

janeyx99 committed:
[skip ci] stable torch library draft

[ghstack-poisoned]

janeyx99 committed:
Update on "[skip ci] stable torch library draft"




[ghstack-poisoned]

janeyx99 committed:
Update on "[skip ci] stable torch library draft"




[ghstack-poisoned]

janeyx99 committed:
Update on "[skip ci] stable torch library draft"




[ghstack-poisoned]

zou3519 commented on caffe2/CMakeLists.txt:
```diff
@@ -503,6 +503,7 @@ endif()
 list(APPEND TORCH_SRCS
   ${TORCH_SRC_DIR}/csrc/autograd/FunctionsManual.cpp
   ${TORCH_SRC_DIR}/csrc/utils/out_types.cpp
+  ${TORCH_SRC_DIR}/library_stable_impl.cpp
```
> Should I add a stable def and fragment when those can be done in python?

C++-only use cases need a def in C++. idk how common it is to have a def in C++ nowadays. If anyone is still using TorchScript, then they would want a def in C++.

janeyx99 replied:
thanks--i've added both

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
- Should I add a stable `def` and `fragment` when those can be done in python?
- Where should library_stable_impl.cpp live?


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
- Where should library_stable_impl.cpp live?


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

albanD commented on torch/library_stable_impl.cpp:
```diff
@@ -0,0 +1,124 @@
+// // in this file, we will implement the stuff in library_stable.h,
```
to delete.

janeyx99 replied:
yep!

albanD commented on buckbuild.bzl:
```diff
@@ -939,6 +939,7 @@ def define_buck_targets(
                 ("", "torch/csrc/**/*.h"),
                 ("", "torch/script.h"),
                 ("", "torch/library.h"),
+                ("", "torch/library_stable.h"),
```
I would prefer a new `stable` subfolder to make this easier to track

janeyx99 replied:
ya this was just i could get some progress in, will figure out the build stuff soon

albanD commented on caffe2/CMakeLists.txt:
```diff
@@ -1259,6 +1260,7 @@ install(FILES
   "${TORCH_SRC_DIR}/extension.h"
   "${TORCH_SRC_DIR}/custom_class.h"
   "${TORCH_SRC_DIR}/library.h"
+  "${TORCH_SRC_DIR}/library_stable.h"
```
You most likely also need to change "buckbuild.bzl" and friends. Looking for "library.h" and pattern matching should work.
Also importing internally just to check might be helpful to avoid noisy revert

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/__init__.py:
```diff
@@ -0,0 +1,26 @@
+import torch
+
+
+so_files = list(Path(__file__).parent.glob("_C*.so"))
+assert len(so_files) == 1, f"Expected one _C*.so file, found {len(so_files)}"
+torch.ops.load_library(so_files[0])
```
```suggestion
import ._C
```

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/__init__.py:
```diff
@@ -0,0 +1,26 @@
+
+from . import ops
+
+# ----------------------------------------------------------------------------- #
+# We've reached the end of what is normal in __init__ files.
+# The following is used to assert the sgd op is properly loaded and
+# calculates correct results upon import of this extension.
+
+param = torch.tensor([1.0, 2.0, 3.0, -0.3, -0.2, -0.1], device="cpu")
+grad = torch.rand_like(param)
+weight_decay = 0.01
+lr = 0.001
+maximize = False
+
+new_param = ops.sgd_out_of_place(param, grad, weight_decay, lr, maximize)
+# expected = param - lr * (grad + weight_decay * param)
+torch._fused_sgd_((param,), (grad,), (), weight_decay=weight_decay, momentum=0.0, lr=lr, dampening=0.0, nesterov=False, maximize=maximize, is_first_step=False)
+assert torch.equal(new_param, param), f"{new_param=}, {param=}"
```
nit: you can have this in a "def test_me()" function that you call from the test file that imports this module.

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/libtorch_agnostic_kernel.cpp:
```diff
@@ -0,0 +1,115 @@
+
+
+// NEED TO DELETE, IS NOT STABLE!!!
+#include <c10/util/Exception.h>
```
This is not used anymore right?

albanD commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/ops.py:
```diff
@@ -0,0 +1,27 @@
+
+
+lib = torch.library.Library("libtorch_agnostic", "FRAGMENT")
+lib.define("sgd_out_of_place(Tensor param, Tensor grad, float weight_decay, float lr, bool maximize) -> Tensor")
```
Are you planning on moving this define to c++ or it will stay here for now?

albanD commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -619,6 +619,22 @@ AOTI_TORCH_EXPORT void aoti_torch_save_tensor_handle(
+AOTI_TORCH_EXPORT AOTITorchError aoti_torch_library_impl(
+    TorchLibraryHandle self,
+    const char* name,
+    void (*fn)(uint64_t*, int64_t, int64_t)
```
nit: if we don't give these types meaningful names, please leave a comment explaining the calling convention.

albanD commented on torch/library_stable.h:
```diff
@@ -0,0 +1,92 @@
+    StableLibrary& operator=(const StableLibrary&) = delete;
+    StableLibrary(StableLibrary&&) = default;
+    StableLibrary& operator=(StableLibrary&&) = default;
+    ~StableLibrary() = default;
```
As discussed offline, we want this to call into something akin to aoti_torch_library_delete

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
 }
+
+// should TorchLibraryOpaque be a uniq_ptr?
+class TorchLibraryOpaque {
```
As discussed offline, this indirection is not needed.

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+      uint64_t *ministack = (uint64_t*)malloc((num_arguments + num_returns) * sizeof(uint64_t));
+      // std::unique_ptr<void *[]> ministack = std::make_unique<void*[]>(num_arguments + num_returns);
+
+      for (size_t idx = 0; idx < num_arguments; idx++) {  // rbarnes will prefer a c10::irange instead of this loop!
```
do it!

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+      for (size_t idx = 0; idx < num_arguments; idx++) {  // rbarnes will prefer a c10::irange instead of this loop!
+        const c10::IValue& arg = torch::jit::peek(stack, idx, num_arguments);
+        if (arg.isInt()) {
+          TORCH_WARN("dealt with an Int");
```
All these warnings need to go?

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+      // schema values, and run the function and modify the uint64_t stack
+      fn_(ministack, num_arguments, num_returns);
+
+      // now pop all inputs on stack. if we pop earlier, Tensors would go out of scope
+      // before calling the function
```
I would not expect this to be true and that all the ATH going through the voidyvoid API are owning.

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+      torch::jit::drop(stack, num_arguments);
+
+      // read the output from the end of the stack and wrap that back into
+      // IValue from void*?
```
```suggestion
      // IValue from uint64_t
```

albanD commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1291,3 +1295,120 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+          AtenTensorHandle ath = torch::aot_inductor::new_tensor_handle(std::move(const_cast<at::Tensor&>(arg.toTensor())));
+          ministack[idx] = from(ath);
+        } else {
+          TORCH_CHECK(false, "Other types of IValues not yet handled!");
```
Once this lands, I would like to do a quick survey of arg and return types for top ops and add more types here accordingly !

albanD commented:
/easycla

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
I wrote this file before learning that there wasn't some cool mechanism to run these tests in CI, so I ended up porting these into one test case in test_cpp_extensions_aot.py.

Questionable whether we should keep this file checked in vs not. Tell me your opinion.

zou3519 replied:
Is this test run in CI? Is there a way to get this test to run in CI?

janeyx99 committed:
Update on "stable torch library draft"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -0,0 +1,77 @@
+        with self.assertRaises(RuntimeError):
+            libtorch_agnostic.ops.identity(
+                t
+            )  # errors as identity shouldn't be registered anymore
```
did you read the lines before this that are super shady :D

albanD commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -623,6 +623,40 @@ AOTI_TORCH_EXPORT void aoti_torch_save_tensor_handle(
+AOTI_TORCH_EXPORT AOTITorchError aoti_torch_library_impl(
+    TorchLibraryHandle self,
+    const char* name,
+    void (*fn)(uint64_t*, int64_t, int64_t));
```
nit: you can move StableIValue definition to this file to use it here

albanD commented on torch/csrc/inductor/aoti_torch/c/shim.h:
```diff
@@ -623,6 +623,40 @@ AOTI_TORCH_EXPORT void aoti_torch_save_tensor_handle(
+AOTI_TORCH_EXPORT AOTITorchError
+aoti_torch_library_def(TorchLibraryHandle self, const char* name);
+
+// Free the torch::Library object
```
nit: add comments to document each of these functions

albanD reviewed (approved):
SGTM
only two small nits

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

zou3519 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1300,3 +1305,138 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
   });
 }
+
+class StableIValueConverter : public c10::OperatorKernel {
```
Why is this a `StableIValueConverter`?

zou3519 replied:
This sounds more like a "StableBoxedKernel"

janeyx99 replied:
This is the boxed struct to contain the functor that liaisons between IValues in torch::jit::stack and StableIValue in the user kernel

janeyx99 replied:
Oh, but the BoxedKernel is the thing that consumes this, no?

zou3519 replied:
My confusion was: this thing seems like a "Kernel", but it's named "StableIValueConverter"

janeyx99 replied:
Renaming to StableIValueBoxedKernel to be super clear cuz it is indeed a kernel that just converts ivalues

zou3519 commented on torch/csrc/inductor/aoti_torch/shim_common.cpp:
```diff
@@ -1300,3 +1305,138 @@ AOTITorchError aoti_torch_cpu__weight_int4pack_mm_cpu_tensor(
+    const auto num_returns = schema.returns().size();
+    const auto num_arguments = schema.arguments().size();
+
+    StableIValue* ministack = (StableIValue*)malloc(
+        (num_arguments + num_returns) * sizeof(StableIValue));
```
nit: Why not std::array? Looks like we're still in C++ land, this buys us some safety (from not needing to free or an exception happening in the middle)

janeyx99 replied:
Using std::vector

zou3519 commented on torch/stable/library.h:
```diff
@@ -0,0 +1,149 @@
+    StableLibrary(
+        Kind kind,
+        const char* ns,
+        const char* k,
```
is this the dispatch key?

janeyx99 replied:
yes

zou3519 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/__init__.py:
```diff
@@ -0,0 +1,22 @@
+with torch._ops.dl_open_guard():
+    # Import the shared library into the process, thus running its
+    # static (global) initialization code in order to register custom
+    # operators with the JIT.
```
"with the JIT" ? you mean the pytorch dispatcher?

janeyx99 replied:
ah this was copied/pasted from the torch.load_library portion, i can just delete this comment since it's confusing

zou3519 commented on test/cpp_extensions/libtorch_agnostic_extension/libtorch_agnostic/csrc/libtorch_agnostic_kernel.cpp:
```diff
@@ -0,0 +1,127 @@
+}
+
+STABLE_TORCH_LIBRARY_IMPL(libtorch_agnostic, CPU, m) {
+  m.impl("libtorch_agnostic::sgd_out_of_place", &boxed_sgd_out_of_place);
```
did you actually need the "libtorch_agnostic::"? (just checking, i think the implementation should be fine)

janeyx99 replied:
nope! i didn't! this was copied from torchao style; it is indeed redundant

zou3519 commented on test/cpp_extensions/libtorch_agnostic_extension/test/test_libtorch_agnostic.py:
```diff
@@ -0,0 +1,78 @@
+
+
+    @onlyCUDA
+    def test_z_delete_torch_lib(self, device):
+        # Why the z + CUDA? THIS TEST MUST BE RUN LAST
+        # We are testing that unloading the library properly deletes the registrations, so running this test
+        # earlier will cause all other tests in this file to fail
```
this probably won't fly as a CI test

janeyx99 replied:
yea lmao, this file's not run in CI, see https://github.com/pytorch/pytorch/pull/148124#discussion_r1985614186

janeyx99 commented on setup.py:
```diff
@@ -1273,6 +1273,7 @@ def main():
         "include/c10/xpu/*.h",
         "include/c10/xpu/impl/*.h",
         "include/torch/*.h",
+        "include/torch/stable/*.h",
```
good call nikita agrees with this

zou3519 replied:
with which lol

janeyx99 replied:
oh lol torch/csrc/stable, he thinks nothing should be in the torch/** namespace

malfet commented on .lintrunner.toml:
```diff
@@ -223,6 +223,7 @@ include_patterns = [
     'c10/**/*.cpp',
     'c10/**/*.h',
     'torch/*.h',
+    'torch/stable/*.h',
```
IMO all C/C++ headers should be in `torch/csrc` folder, also `stable` is, IMO, confusing name. How about `capi`?

janeyx99 replied:
The C API is the shim.h in aoti/c/, this directory is meant to host the C++ header only files that will only call into the shim.h API or be self contained.

malfet commented on torch/stable/library.h:
```diff
@@ -0,0 +1,149 @@
+#include <torch/csrc/inductor/aoti_torch/c/shim.h>
+
+
+template <typename T>
+StableIValue from(T val) {
+  static_assert(
+      sizeof(T) <= sizeof(StableIValue),
+      "StableLibrary stack does not support parameter types larger than 64 bits.");
+  return *reinterpret_cast<StableIValue*>(&val);
+}
+
+template <typename T>
+T to(StableIValue val) {
+  return *reinterpret_cast<T*>(&val);
+}
```
It looks cool, but templates like that are by design unstable, as they will produce different symbols depending on whether or not you compile your code with `-frtti` vs `-fno-rtti` , `-fexceptions`/`-fnoexceptions` and perhaps so on. If you really want a stable API(and ABI i.e. compile once use across versions of torch) you'll need C api, or something that explicitly decorated as `nothrow` (though this has to be checked)

malfet commented on torch/stable/library.h:
```diff
@@ -0,0 +1,149 @@
+  private:
+    TorchLibraryHandle lib_;
+  public:
+    enum Kind {
```
No C-style enums in 2025 please
```suggestion
    enum class Kind {
```

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

Things to resolve:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant


- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

malfet commented:
@pytorchbot revert -m "Sorry for reverting your PR, but somehow it caused test failures in newly introduced tests, see https://hud.pytorch.org/hud/pytorch/pytorch/main/1?per_page=50&name_filter=pull%20%2F%20linux-focal-cuda12.6-py3.10-gcc11-sm89%20%2F%20test%20(default%2C%201&mergeLF=true" -c weird

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 commented on test/test_cpp_extensions_aot.py:
```diff
@@ -227,6 +227,49 @@ def test_python_agnostic(self):
+            _run_identity(init_mem, device)
+            curr_mem = torch.cuda.memory_allocated(device)
+            self.assertEqual(curr_mem, init_mem)
+
```
Ended up removing the third test as it is undefined behavior sometimes what happens when u try to call an unregistered op

janeyx99 commented:
@pytorchbot merge

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]

janeyx99 commented:
@pytorchbot merge

janeyx99 commented:
@pytorchbot merge

albanD commented:
@pytorchbot merge -r

janeyx99 committed:
Update

[ghstack-poisoned]

janeyx99 committed:
Update on "Add a stable TORCH_LIBRARY to C shim"

This PR adds two main parts:
- shim.h stable C APIs into torch::Library APIs
- a higher level API in torch/csrc/stable/library.h that calls into this shim.h + otherwise is self contained

Goal: custom kernel writers should be able to call the apis in the directories above in order to register their library in a way that allows their custom extension to run with a different libtorch version than it was built with.

Subplots resolved:

- Do we want a whole separate StableLibrary or do we want to freeze torch::Library and add `m.stable_impl(cstring, void (*fn)(void **, int64_t, int64_t)` into it
    - Yes, we want a separate StableLibrary. We cannot freeze Library and it is NOT header only.
- Should I use unint64_t as the common denominator instead of void* to support 32bit architectures better?
    -  Yes, and done
- Should I add a stable `def` and `fragment` when those can be done in python?
    - I think we do want these --- and now they're done
- Where should library_stable_impl.cpp live? -- no longer relevant
- I need some solid test cases to make sure everything's going ok. I've intentionally thrown in a bunch of random dtypes into the signature, but I still haven't tested returning multiple things, returning nothing, complex dtypes, etc.
    - Have since tested all the torch library endpoints. the others can be tested in a followup to separate components that need to be in shim.h vs can be added later




[ghstack-poisoned]
